source/shared/cpp/ObjectModel/EnumMagic.h

```diff
@@ -0,0 +1,128 @@
+        const EnumMapping<ENUMTYPE>& get##ENUMTYPE##Enum(); \
+    }                                                       \
+    const std::string ENUMTYPE##ToString(const ENUMTYPE t); \
+    ENUMTYPE ENUMTYPE##FromString(const std::string& t);
+
+#define _DEFINE_ADAPTIVECARD_ENUM_INVARIANT(ENUMTYPE, ...)                \
+    namespace EnumHelpers                                                 \
+    {                                                                     \
+        const EnumMapping<ENUMTYPE>& get##ENUMTYPE##Enum()                \
+        {                                                                 \
+            static const EnumMapping<ENUMTYPE> generatedEnum __VA_ARGS__; \
+            return generatedEnum;                                         \
+        }                                                                 \
+    }                                                                     \
+    const std::string ENUMTYPE##ToString(const ENUMTYPE t) { return EnumHelpers::get##ENUMTYPE##Enum().toString(t); }
+
+// Define mapping functions for ENUMTYPE. Throw an exception if caller passes in a string that doesn't map
+#define DEFINE_ADAPTIVECARD_ENUM_THROW(ENUMTYPE, ...)                \
+    _DEFINE_ADAPTIVECARD_ENUM_INVARIANT(ENUMTYPE, __VA_ARGS__);      \
+    ENUMTYPE ENUMTYPE##FromString(const std::string& t)              \
+    {                                                                \
+        try                                                          \
+        {                                                            \
+            return EnumHelpers::get##ENUMTYPE##Enum().fromString(t); \
+        }                                                            \
+        catch (const std::out_of_range&)                             \
+        {                                                            \
+            throw std::out_of_range("Invalid " #ENUMTYPE ": " + t);  \
+        }                                                            \
+    }
+
+// Define mapping from string function for ENUMTYPE. Use the specified DEFAULT when a caller passes in a string that
+// doesn't map.
+#define DEFINE_ADAPTIVECARD_ENUM_DEFAULT(ENUMTYPE, DEFAULT, ...)     \
+    _DEFINE_ADAPTIVECARD_ENUM_INVARIANT(ENUMTYPE, __VA_ARGS__);      \
+    ENUMTYPE ENUMTYPE##FromString(const std::string& t)              \
+    {                                                                \
+        try                                                          \
+        {                                                            \
+            return EnumHelpers::get##ENUMTYPE##Enum().fromString(t); \
+        }                                                            \
+        catch (const std::out_of_range&)                             \
+        {                                                            \
+            return DEFAULT;                                          \
```

Comment:
is DEFAULT needed to be verified that it is correct type?

Reply:
The build will fail if it's invalid

---
In reply to: [231267703](https://github.com/Microsoft/AdaptiveCards/pull/2151#discussion_r231267703) [](ancestors = 231267703)